### PR TITLE
Respect VAULT_SKIP_VERIFY envionment variable setting in hashi_vault lookup plugin

### PR DESCRIPTION
**Pull request by markafarrell**
##### SUMMARY
The standard way of disabling certificate verification for hashicorp vault is by setting the `VAULT_SKIP_VERIFY` environment variable to True.

Currently the hashi_vault lookup plugin does not respect the setting of this environment variable (you instead need to set `validate_certs` to `False` to disable certificate verification).

This adds an the functionality to set the value of `validate_certs` to the inverse of `VAULT_SKIP_VERIFY` (if it is set)

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

hashi_vault

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: plugins/lookup/hashi_vault.py
Comment (felixfontein) [inline plugins/lookup/hashi_vault.py:127]:
```suggestion
          version_added: 1.3.0
```
Patch releases do not get new features (https://semver.org/).

**Review by felixfontein — COMMENTED** [review]
I'm not sure whether adding *another* option for validation is a good idea. It makes everything more complicated.

Maybe it's better to remove the explicit default for `validate_certs`, and handle the env fallback manually in the code.

**Comment by felixfontein** [comment]
CC @briantist

**Comment by markafarrell** [comment]
> I'm not sure whether adding _another_ option for validation is a good idea. It makes everything more complicated.
> 
> Maybe it's better to remove the explicit default for `validate_certs`, and handle the env fallback manually in the code.

I wasn't sure which was better either. Happy to got down the `validate_certs` that gets populated manually with `not VAULT_SKIP_VERIFY` if you would prefer

**Comment by briantist** [comment]
Right now main issue I see with this is that the env var overrides the plugin param. An explicit parameter setting should take precedence over the env.

To do that, while still preserving previous behavior, will require _removing_ the currently set `default: True` so that you can test whether a value was provided, and then, if the env isn't set and no value for the param was provided, you can then manually set the default value to `true`.

**Comment by felixfontein** [comment]
In any case, this also needs a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to).

**Comment by briantist** [comment]
I have another possible suggestion, that's more involved but possibly better long term:
1. _Do_ add a new parameter `skip_cert_validation` which uses the built-in env processing to consume the `VAULT_SKIP_VERIFY` env var, and defaults to `False`
1. Deprecate `validate_certs` to some future version, following the standard deprecation rules for this project
1. Explain in detail in the docs that `skip_cert_validation` will be taking the place of `validate_certs` and that if both values are specified, then `skip_cert_validation` takes precedence.

The idea being that this gives backward compatibility for existing playbooks over the deprecation period, while allowing those aware of the upcoming changes to choose a forward compatible option during that same deprecation period.

Devils in the details are:
* You still need to change how defaults are specified/interpreted during the deprecation period, on at least one of the params, so that you can know when one they are explicitly specified and apply the deprecation period precedence rules.
* That may require a change after deprecation that restores more standard param processing (but wouldn't change behavior).

---

This sounds more complicated but trying to give a well-supported avenue for making a forward-looking change that better matches the way vault operates.

This is just off the top of my head so I may very well have overlooked something important; going to lean a little bit on @felixfontein who has a really good sense for changes like this :)

**Comment by markafarrell** [comment]
> This sounds more complicated but trying to give a well-supported avenue for making a forward-looking change that better matches the way vault operates.

This was my original intention but handling the deprecation of the old option seemed like it would be a hassle.

I'm not sure the extra effort really provides any benefit (to either developer or user of the lookup). 

I will make the changes to the precedence you suggested above: (I think this will be needed regardless)

- Explicit setting of `validate_certs`
- Explicit setting of `VAULT_SKIP_VERIFY`
- Default to `True`

Then ill wait for input form @felixfontein and yourself @briantist on which route you would like to take

**Comment by markafarrell** [comment]
On a side note. 

When I attempt to run `ansible-test integration --docker centos8 -v lookup_hashi_vault` the tests fail due to the `community.crypto` collection not being installed.

I'm not totally sure how `ansible-test` works but should `ansible-galaxy -vvv collection install community.crypto` be added to `tests/integration/targets/lookup_hashi_vault/runme.sh` or is there somewhere else that should trigger the install of the the required collections?

**Comment by felixfontein** [comment]
I think it's better to stick to how things are in Ansible, and there the option is always called `validate_certs`. It's really confusing IMO to replace it by an inverted option just in this case, and keep `validate_certs` everywhere else.

@markafarrell about the community.crypto dependency: you need to install collection dependencies **before** running `ansible-test`. I think you even need to install them *next* to the collection you're testing, so if community.general is in `/path/to/ansible_collections/community/general`, then community.crypto should be installed at `/path/to/ansible_collections/community/crypto`.

File: plugins/lookup/hashi_vault.py
Comment (felixfontein) [inline plugins/lookup/hashi_vault.py:121]:
```suggestion
        - Will be populated with the inverse of C(VAULT_SKIP_VERIFY) if that is set and `validate_certs` is not explicitly provided (added in community.general 1.3.0).
```

File: plugins/lookup/hashi_vault.py
Comment (felixfontein) [inline plugins/lookup/hashi_vault.py:122]:
```suggestion
        - Will default to C(true) if neither I(validate_certs) or C(VAULT_SKIP_VERIFY) are set.
```

File: tests/integration/targets/lookup_hashi_vault/lookup_hashi_vault/tasks/approle_test.yml
Comment (felixfontein) [inline tests/integration/targets/lookup_hashi_vault/lookup_hashi_vault/tasks/approle_test.yml:33]: Good spot, right now the failure happens because of the misspelling of the lookup name ;)

**Comment by markafarrell** [comment]
@felixfontein Ok. I would have thought at dependencies would have been been installed for you. 
Weirdly this (https://github.com/ansible-collections/community.general/blob/main/tests/requirements.yml) exists but I'm not sure how or where it is used.

**Comment by felixfontein** [comment]
That file isn't used yet. I guess a future version of ansible-test will/might support it, but right now it's not supported.

File: tests/integration/targets/lookup_hashi_vault/lookup_hashi_vault/tasks/approle_test.yml
Comment (markafarrell) [reply tests/integration/targets/lookup_hashi_vault/lookup_hashi_vault/tasks/approle_test.yml:33]: One additional thing that caught me out was requiring a space on the end of `conn_params`. We should probably:
- put `conn_params` at the end or have a space before 'secret='

**Comment by felixfontein** [comment]
(The dependencies are currently installed explicitly in https://github.com/ansible-collections/community.general/blob/main/tests/utils/shippable/shippable.sh#L70-L75)

File: tests/integration/targets/lookup_hashi_vault/lookup_hashi_vault/tasks/approle_test.yml
Comment (briantist) [reply tests/integration/targets/lookup_hashi_vault/lookup_hashi_vault/tasks/approle_test.yml:33]: Do you think adding a comment near where `conn_params` is defined pointing out the importance of the trailing space would be enough?

**Comment by briantist** [comment]
> I think it's better to stick to how things are in Ansible, and there the option is always called `validate_certs`. It's really confusing IMO to replace it by an inverted option just in this case, and keep `validate_certs` everywhere else.

Thanks! This is the kind of input of I was hoping for, as I wasn't so aware that `validate_certs` had such a level of commonality across Ansible

File: plugins/lookup/hashi_vault.py
Comment (felixfontein) [inline plugins/lookup/hashi_vault.py:121]:
```suggestion
        - Will be populated with the inverse of C(VAULT_SKIP_VERIFY) if that is set and I(validate_certs) is not explicitly
```

File: plugins/lookup/hashi_vault.py
Comment (felixfontein) [inline plugins/lookup/hashi_vault.py:122]:
```suggestion
          provided (added in community.general 1.3.0).
```
This continues the paragraph in the previous list element, and doesn't start a new one.

File: tests/integration/targets/lookup_hashi_vault/lookup_hashi_vault/tasks/approle_test.yml
Comment (felixfontein) [reply tests/integration/targets/lookup_hashi_vault/lookup_hashi_vault/tasks/approle_test.yml:33]: Adding a comment in this PR is ok, but adjusting all use-cases of `conn_params` should be better done in another PR (if you want to do this).

**Comment by markafarrell** [comment]
@felixfontein I will open another PR to fix the whitespace problem.

File: tests/integration/targets/lookup_hashi_vault/lookup_hashi_vault/tasks/approle_test.yml
Comment (markafarrell) [reply tests/integration/targets/lookup_hashi_vault/lookup_hashi_vault/tasks/approle_test.yml:33]: #1035  opened to address this

**Comment by markafarrell** [comment]
@felixfontein @briantist Is there anything more required to be done here?

**Comment by felixfontein** [comment]
@markafarrell to me the PR looks good, though I'm not maintaining this plugin. Once @briantist thinks its fine to merge I'll merge it.

**Review by briantist — APPROVED** [review]
The changes look good to me, thanks @markafarrell ! I'll also look at #1035 and comment there separately.

I have one suggestion that I'll leave to @felixfontein 's judgement whether it should prevent merging, but it would be good to add a test for when `VAULT_SKIP_VERIFY` is set to something that fails to convert to boolean (to prevent future regressions in that exception path).

**Comment by felixfontein** [comment]
@briantist that's a bit harder to check: you would need to set up an instance with a certificate that does not validate, and then make sure that in case `VAULT_SKIP_VERIFY` is set to something that fails to convert, the lookup actually fails. I don't really know the hashi_vault tests; if such a variant already exists and can be used here, I think this should be tested, otherwise it's too much work for this PR and is better suited for a follow-up (if you really want this).

**Comment by briantist** [comment]
> @briantist that's a bit harder to check: you would need to set up an instance with a certificate that does not validate, and then make sure that in case `VAULT_SKIP_VERIFY` is set to something that fails to convert, the lookup actually fails. I don't really know the hashi_vault tests; if such a variant already exists and can be used here, I think this should be tested, otherwise it's too much work for this PR and is better suited for a follow-up (if you really want this).

You know @felixfontein , that's completely right 🤦‍♂️ I was thinking about it like we could just check the value the parameter ends up with but I forgot these are not unit tests. I'm a little out of it this week.

Sorry to hold this up any longer, looking forward to it being merged. Thanks again @markafarrell !

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@markafarrell thanks for implementing this!
@briantist thanks for reviewing this!

**Pull request merged**
